crates/paymaster-execution/src/diagnostics/extractors/avnu.rs

```diff
@@ -0,0 +1,1135 @@
+
+    /// Token amount is zero
+    pub const TOKEN_AMOUNT_ZERO: &str = "token from amount is 0";
+
+    /// Routes array is empty
+    pub const ROUTES_EMPTY: &str = "routes is empty";
+
+    /// First route sell token doesn't match
+    pub const INVALID_TOKEN_FROM: &str = "invalid token from";
+
+    /// Last route buy token doesn't match
+    pub const INVALID_TOKEN_TO: &str = "invalid token to";
+
+    /// Unknown exchange in routes
+    pub const UNKNOWN_EXCHANGE: &str = "unknown exchange";
+}
+
+/// Selectors for AVNU Exchange functions.
+mod selectors {
+    use starknet::core::types::Felt;
+    use starknet::macros::selector;
+
+    pub fn multi_route_swap() -> Felt {
+        selector!("multi_route_swap")
+    }
+
+    pub fn swap_exact_token_to() -> Felt {
+        selector!("swap_exact_token_to")
+    }
+
+    pub fn swap_external_solver() -> Felt {
+        selector!("swap_external_solver")
+    }
+}
+
+/// Extractor for AVNU Exchange contract errors.
+///
+/// Handles the following swap functions:
+/// - `multi_route_swap`: Standard swap with slippage protection
+/// - `swap_exact_token_to`: Swap to receive exact amount of buy token
+/// - `swap_external_solver`: Swap using external solver
+///
+/// # Extracted Metadata
+///
+/// For swap errors, extracts:
+/// - `sell_token`: Token being sold (address)
+/// - `sell_token_symbol`: Token symbol (e.g., "ETH") - when available
+/// - `sell_token_name`: Token name (e.g., "Ethereum") - when available
+/// - `sell_amount`: Normalized amount of sell token - when token info available
+/// - `sell_amount_hex`: Raw amount as hex string (always present)
+/// - `buy_token`: Token being bought (address)
+/// - `buy_token_symbol`: Token symbol - when available
+/// - `buy_token_name`: Token name - when available
+/// - `buy_amount`: Normalized amount - when token info available
+/// - `buy_amount_hex`: Raw amount as hex string (always present)
+/// - `buy_min_amount` / `buy_min_amount_hex`: Minimum expected (for slippage)
+/// - `beneficiary`: Recipient of the swap
+#[derive(Clone)]
+pub struct AvnuExtractor {
+    /// The AVNU Exchange contract address
+    contract_address: Felt,
+    /// Optional token service for enriching metadata
+    token_service: Option<TokenService>,
+}
+
+impl AvnuExtractor {
+    /// Creates a new AVNU extractor for the given contract address.
+    pub fn new(contract_address: Felt) -> Self {
+        Self {
+            contract_address,
+            token_service: None,
+        }
+    }
+
+    /// Creates a new AVNU extractor with a token service for enriched metadata.
+    pub fn with_token_service(contract_address: Felt, token_service: TokenService) -> Self {
+        Self {
+            contract_address,
+            token_service: Some(token_service),
+        }
+    }
+
+    /// Checks if the error message indicates a slippage error.
+    fn is_slippage_error(&self, error: &str) -> bool {
```

Comment:
Why not have a proper "Error" type that has "new(error: &str)" and produce an error category ? It seems weird to have that here ?